src/psyclone/psyir/transformations/debug_checksum_trans.py

```diff
@@ -0,0 +1,203 @@
+    ... DebugChecksumTrans().apply(loop)
+    ... print(FortranWriter()(psyir))
+    subroutine mysubroutine()
+      integer, dimension(10,10) :: a
+      integer :: i
+      integer :: j
+      integer :: PSYCLONE_INTERNAL_line_
+    <BLANKLINE>
+      do i = 1, 10, 1
+        do j = 1, 10, 1
+          a(i,j) = a(i,j) + i - j
+        enddo
+      enddo
+      PSYCLONE_INTERNAL_line_ = __LINE__
+      PRINT *, "PSyclone checksums from mysubroutine at line:", \
+PSYCLONE_INTERNAL_line_ + 1
+      PRINT *, "a checksum", SUM(a)
+    <BLANKLINE>
+    end subroutine mysubroutine
+    <BLANKLINE>
+
+    '''
+    def apply(self, node: Union[Node, List[Node]], options=None) -> None:
+        '''
+        Applies the checksum transformation to the provided node(s).
+
+        :param node: The node or list of nodes to apply the
+                      transformation to.
+        :param options: a dictionary with options for transformations.
+        :type options: Optional[Dict[str, Any]]
+
+        '''
+        self.validate(node, options={"node-type-check": False})
+
+        node_list = self.get_node_list(node)
+        routine = node_list[0].ancestor(Routine)
+        routine_table = routine.symbol_table
+
+        fwriter = FortranWriter()
+        writes = []
+        # Loop over the assignments in the region
+        assigns = []
+        for node in node_list:
+            assigns.extend(node.walk(Assignment))
+        # Loop through the assignments and find the arrays
+        for assign in assigns:
+            if isinstance(assign.lhs, StructureAccessorMixin):
+                # Find the last member.
+                member = assign.lhs.member
+                while isinstance(member, StructureAccessorMixin):
+                    member = member.member
+                datatype = assign.lhs.datatype
+                while not isinstance(datatype, ScalarType):
+                    datatype = datatype.datatype
+                if (member.is_array and datatype.intrinsic in
+                        [ScalarType.Intrinsic.REAL,
+                         ScalarType.Intrinsic.INTEGER]):
+                    writes.append(assign.lhs)
+            elif (assign.lhs.is_array and assign.lhs.datatype.intrinsic in
+                  [ScalarType.Intrinsic.REAL, ScalarType.Intrinsic.INTEGER]):
+                writes.append(assign.lhs)
+
+        # For each write, add a checksum after.
+        checksum_nodes = []
+        freader = FortranReader()
+        for lhs in writes:
+            # Need to convert the lhs to a full range variant.
+            copy = lhs.copy()
+            name, _ = copy.get_signature_and_indices()
+            if isinstance(lhs, StructureAccessorMixin):
+                member = copy.member
+                while isinstance(member, StructureAccessorMixin):
+                    member = member.member
+                datatype = assign.lhs.datatype
+                for i in range(len(member.indices)):
```

Comment:
Unfortunately, I think this is assuming that you always have, e.g. `myvar%something%else(i,j) = ...` while you could well have `myvar%something(i,j)%else = ...`. Rather than get bogged down in this, I suggest we just exclude structure references for now. Alternatively, you could look at using VariablesAccessInfo?

Reply:
Actually, rather than exclude all types, we could just exclude those that this code currently doesn't work for (e.g. `myvar%something(i,j)%else`).